More thorough aliasing detection in nonscalar copy, broadcast and assignment

mbauman opened the pull request:
This commit puts together the beginnings of an "interface" for detecting and avoiding aliasing between arrays before performing some mutating operation on one of them.

`A′ = unalias(dest, A)` is the main entry point; it checks to see if `dest` and `A` might alias one another, and then either returns `A` or a copy of `A`. When it returns a copy of `A`, it absolutely must preserve the same type. As such, this function ~~is designed to be specialized on `A`, typically like:~~ calls an internal ~~`copypreservingtype`~~ `unaliascopy` function, which defaults to `copy` but ensures it returns the same type.

`mightalias(A, B)` is a quick and ~~rough~~ conservative check to see if two arrays alias eachother.  Instead of requiring every array to know about every other array type, it simply asks both arrays for their `dataids` and then looks for an empty intersection.

Finally, `dataids(A)` returns a tuple of ~~ranges~~ `UInt`s that describe something about the region of memory the array occupies. ~~It could be simply `(objectid(A):objectid(A),)`~~ It defaults to simply `(objectid(A),)`, or for an array with multiple fields, `(dataids(A.a)..., dataids(A.b)..., ...)`.

Fixes #21693, fixes #15209, and a pre-req to get tests passing for #24368 (since we had spot checks for `Array` aliasing, but moving to broadcast means we need a slightly more generic solution).  I haven't exported any of these functions since I'd like us to use them a bit more before we officially document and support them for all custom arrays.

mbauman committed:
More thorough aliasing detection in nonscalar copy, broadcast and assignment.

This commit puts together the beginnings of an "interface" for detecting and avoiding aliasing between arrays before performing some mutating operation on one of them.

`A′ = unalias(dest, A)` is the main entry point; it checks to see if `dest` and `A` might alias one another, and then either returns `A` or a copy of `A`. When it returns a copy of `A`, it absolutely must preserve the same type. As such, this function is designed to be specialized on `A`, typically like:

    Base.unalias(dest, A::T) = mightalias(dest, A) ? typeof(A)(unalias(dest, A.a), unalias(dest, A.b), ...) : A

`mightalias(A, B)` is a quick and rough check to see if two arrays alias eachother.  Instead of requiring every array to know about every other array type, it simply asks both arrays for their `dataids` and then looks for an empty intersection between every pair.

Finally, `dataids(A)` returns a tuple of ranges that describe something about the region of memory the array occupies. It could be simply `(objectid(A):objectid(A),)` or for an array with multiple fields, `(dataids(A.a)..., dataids(A.b)..., ...)`.

nalimilan commented on base/abstractarray.jl:
```diff
@@ -1048,6 +1048,50 @@ function _setindex!(::IndexCartesian, A::AbstractArray, v, I::Vararg{Int,M}) whe
+"""
+    unalias(dest, A)
+
+Return either A or a copy of A, in a rough effort to prevent modifications to `dest` from
```
Missing backticks around `A`.

nalimilan commented on base/abstractarray.jl:
```diff
@@ -1048,6 +1048,50 @@ function _setindex!(::IndexCartesian, A::AbstractArray, v, I::Vararg{Int,M}) whe
+
+This function must return an object of exactly the same type as A for performance and type-stability.
+
+See also `mightalias` and `dataids`.
```
Use `@ref`.

nalimilan commented on base/abstractarray.jl:
```diff
@@ -1048,6 +1048,50 @@ function _setindex!(::IndexCartesian, A::AbstractArray, v, I::Vararg{Int,M}) whe
+"""
+unalias(dest, A) = A
+unalias(dest, A::Array) = mightalias(dest, A) ? copy(A) : A
+@inline unalias(dest, As::Tuple) = (unalias(dest, As[1]), unalias(dest, tail(As))...)
```
Why not use varargs?

mbauman replied:
Why not?  I liked this symmetry in the calling conventions:

```
function setindex(dest, source, I...)
    source′ = unalias(dest, source)
    I′ = unalias(dest, I)
```

It's also nice in that the tuple case mostly fits the explanation above without needing to call it out — it'll either return exactly the tuple or a (partial) copy.

nalimilan replied:
Yeah, I guess you need to choose between symmetry and consistency with `setindex`.

nalimilan commented on base/abstractarray.jl:
```diff
@@ -1048,6 +1048,50 @@ function _setindex!(::IndexCartesian, A::AbstractArray, v, I::Vararg{Int,M}) whe
+
+Perform a rudimentary test to check if arrays A and B might share the same memory.
+
+Defaults to false unless `dataids` is specialized for both `A` and `B`.
```
When in doubt, wouldn't it be better to return `true`? More generally, shouldn't the functions require/guarantee that any aliasing is detected and that in doubt a copy will be made? The use of terms like "rough" and "rudimentary" doesn't inspire confidence. ;-) We don't want silent corruption to happen by default...

Or it frequently impossible to determine whether arrays alias?

mbauman replied:
Indeed, I don't want to inspire confidence.  That's why this isn't exported.  I just want to do incrementally better than we had been doing before for the builtin array types.  There's an asymmetry in costs here — returning `true` by default would lead to _lots_ of unnecessary copies.  But we also need to perform this check quite frequently, so I don't want the check itself to be expensive.  This is the tradeoff I came to.

I initially had default fallbacks for strided arrays based on `pointer(A, firstindex(A))` and `pointer(A, lastindex(A))` and all other arrays based on `objectid(A)` — that led to #25858.  We could try to extend this more generically later.

mbauman commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25890#issuecomment-363126709) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/5e1f5160696bcedf26b601073118c4ef11ad246b/55cdfc9_vs_842ff17/report.md). cc @ararslan

mbauman committed:
Skip aliasing checks in one-argument broadcast!

mbauman committed:
Fixup docstrings from review

mbauman committed:
Add support for dealiasing Adjoint and Transpose

And fixup the assumption that `broadcast!(f, C, A)` is safe

mbauman commented:
Thanks Nanosoldier.  It had flagged bonus copies coming from `broadcast!(conj, A, A)`.  This is indeed safe — we only make one pass through the array… but it's only safe because the destination `===` the source.

mbauman commented:
Any further thoughts here?  I think this infrastructure gives us a nice place to expand this functionality in the future.  The biggest question I had in implementing it is if we should try harder to detect aliasing in the generic case.  I had initially had these definitions:

```julia
unalias(dest, A::AbstractArray) = mightalias(dest, A) ? deepcopy(A) : A
dataids(A::AbstractArray) = A === parent(A) ? (objectid(A):objectid(A),) : dataids(parent(A))
_increasingrange(a, b) = min(a,b):max(a,b)
dataids(A::StridedArray) = (_increasingrange(UInt(pointer(A, firstindex(A))), UInt(pointer(A, lastindex(A)))),)
```

But we could add these _at any point_.  In fact, we could eventually lean on the GC or more private struct construction internals to recursively identify and dealias fields if we wished.

JeffBezanson commented on base/abstractarray.jl:
```diff
@@ -1048,6 +1048,50 @@ function _setindex!(::IndexCartesian, A::AbstractArray, v, I::Vararg{Int,M}) whe
+"""
+unalias(dest, A) = A
+unalias(dest, A::Array) = mightalias(dest, A) ? copy(A) : A
+@inline unalias(dest, As::Tuple) = (unalias(dest, As[1]), unalias(dest, tail(As))...)
```
I think this would make more sense as a varargs function; otherwise it seems like you're unaliasing the tuple itself (which is somewhat array-like, so could be confusing).

mbauman replied:
Yeah, that was Milan's comment, too.  I can make that change.  https://github.com/JuliaLang/julia/pull/25890#discussion_r166006058

mbauman commented:
I'll plan to merge this tomorrow unless I hear any objections.

vtjnash commented on base/array.jl:
```diff
@@ -694,22 +694,18 @@ function setindex! end
     @_propagate_inbounds_meta
-    A === I && (I = copy(I))
-    for i in I
+    J = unalias(A, I)
```
Since we were just calling `copy` before, maybe we should just keep doing that, and just improve the `A === I` test. We won't really need the `unalias` function then (which looks rather complex to implement), and this would just be written `mayalias(A, I) && (I = copy(I))`

I assume that in cases, the code might also just instead branch to an alternative implementation, so the copy isn't always necessary, just a defensive algorithm?

mbauman replied:
This is a slightly more special case, since we know that `A::Array`, so if `A === I` then `I::Array`, too.  And so we know that `copy(I)::typeof(I)`.  Note that `unalias(dest, A::Array)` is exactly this definition — it just calls `copy(A)`.

In the general case, a branch with `copy(I)` or `I` will be type-unstable, necessitating loop switching or a function barrier or some such.  That's why custom arrays need to implement it on their own.

Sure, you'd be free to call `mightalias` instead of `unalias` if you know about an alternative.

nalimilan commented:
I still (see https://github.com/JuliaLang/julia/pull/25890#discussion_r166006989) find it unsatisfying that `might_alias` does even guarantee that there will be no false negatives, meaning that this general mechanism may sometimes fail to prevent corruption. I'd rather make a copy or raise an error in that case by default (as for sparse matrices, https://github.com/JuliaLang/julia/issues/21693). But if that's just me and there's no efficient solution to that, go ahead...

mbauman committed:
Refactor API to use `map(x->unalias(dest, x), xs)` instead of...

a builtin mapper method. Using varargs would require de-structuring the one arg case: `(a,) = unalias(dest, a)`.

mbauman committed:
fixup! missing sparse transition to new API

mbauman commented:
So as a compromise I tried implementing the slightly better fallback methods I described in https://github.com/JuliaLang/julia/pull/25890#issuecomment-363843095, but I ran into performance issues since `dataids(A::AbstractArray) = A === parent(A) ? (objectid(A):objectid(A),) : dataids(parent(A))` cannot inline and isn't type-stable.  I also noticed that `deepcopy` inlines at least a bit of its work into the caller, which adds quite a bit of unnecessary code weight — but that has an easier solution.

One reason not to do the `dataids(::Any) = (UInt(0):typemax(UInt),)` is because that'd necessitate us documenting, exporting, and standing by this interface.  Doing it incrementally like this means that we can test it in base and then roll it out to all custom arrays as a new feature in 1.x.

mbauman committed:
Add support for BitArray

mbauman commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25890#issuecomment-365454824) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/4ac1bd80ba7925cc774da8084105b3eca13705a4/c37e21c_vs_160a467/report.md). cc @ararslan

mbauman commented:
The performance regressions in `lucompletepivSub` are actually _bug fixes_.  That code [attempts to swap columns and rows](https://github.com/JuliaCI/BaseBenchmarks.jl/blob/efb9b36187ae3257c1390e783535347430ad1e7a/src/array/subarray.jl#L37) with an idiom like: `A[1:end,[1,2]] = @view A[1:end,[2,1]]`.  This isn't safe without this patch.

`laplace_iter_sub` similarly does [clever assignments into a view from views](https://github.com/JuliaCI/BaseBenchmarks.jl/blob/f6dfcc0c8b45d6d3c43f94548994494ae2c69427/src/problem/Laplacian.jl#L97-L103) of the same array.  I _think_ that this one is done in a way that happens to be safe, but it does accumulate `eps`-level differences with more than 14 iterations when I add a `copy` around the RHS of the `.=`.  Not sure what's happening there, but this perf change is not surprising.

`euro_option_vec` is doing [`S .*= exp.(t2 .* randn!(R) .+ t1)`](https://github.com/JuliaCI/BaseBenchmarks.jl/blob/2b9da681941a1794d99f700d3e9aca3fadb9c863/src/problem/MonteCarlo.jl#L44), which is effectively `broadcast!(f, S, S, t2, randn!(R), t1)`, so this, too, fails the aliasing check.  We're just being a little too strict here — had it been `broadcast!(f, S, S', …) ` (or some other such re-arranging view), then it would have been a bug fix.   `nbody` is doing the same thing.  I think we should be able to extend my one-argument perf patch (94262bdda739d234a35fe8a4a62f11d7547815f5) to any arg.

`repeat` is a slightly different problem — it's "tiling" the repeated region from the top-left corner throughout the array.  I should be able to fix this for `StridedArray`s by adding a specialization for `StridedSubArray`s that looks at the pointers of the restricted region.

mbauman committed:
Extend broadcast's `===` aliasing exemption to varargs

mbauman committed:
Add default StridedArray dataids definition and constrain StridedSubArray...

to only look at the linear data segment of its parent that it can reach

mbauman commented:
Ok, I've added two commits that try to ameliorate those performance concerns (2254dc5 and d97d0c3 — GitHub is showing them out of order due to a rebase).

Unfortunately, d97d0c3 isn't sufficient to fix the performance regression in `repeat`, for two reasons: first, it's using the `A[I] = B` syntax, which doesn't consider indices in its computation — but that's slated to change to broadcast, which resolves this part of the issue.  The bigger problem is that I'm only looking at linear segments of memory, and while doing an operation like an N-dimensional tile is strided, I'm only looking at the linear maximal extents of the memory.

It feels rather unsatisfactory, but I think I'll need to change `repeat` to copy the prototype to a temporary buffer and then tile that throughout the output array.

StefanKarpinski commented:
Can we just have a way of expressing "just trust me, don't worry about aliasing here"?

mbauman committed:
More refined aliasing detection for views into the same parent

And a more accurate StridedArray dataids definition.

mbauman commented:
I feel like this is just something we need to do better at.  There are lots of things in this vein — we've wanted to include dimensionality checks in a `@boundscheck` block before but decided that it was surprising.

I started porting numpy's [implementation of a solver of bounded Diophantine equations for checking strided array overlaps (BSD3)](https://github.com/numpy/numpy/blob/5acb9b3c18cb8001ef30635b7aae69fb12152395/numpy/core/src/private/mem_overlap.c) to Julia before realizing overnight that the most common case in Julia is _so much simpler_.  The case in `repeat` — and, I believe, most commonly — is that you're assigning two views into the same parent with indices that do not overlap. Since we hold onto our indices, we can just sequentially check each dimension for overlaps.  We can always further refine this in the future.

I was also inspired by numpy's implementation to correct and improve the default `dataids(::StridedArray)` implementation.  It's a simple algorithm but is definitely inspired by their work — should I add a `Copyright (c) 2015 Pauli Virtanen` near the implementation?

So the status now is that `Base.mightalias(::StridedArray, ::StridedArray)` will give you a correct and conservative answer.  `Base.mightalias(::AbstractArray, ::AbstractArray)` might lie to you if one of the array implementations hasn't defined their `dataids`.

vtjnash commented:
> It's a simple algorithm but is definitely inspired by their work — should I add a Copyright (c) 2015 Pauli Virtanen near the implementation?

Algorithms can't be copyright (at least in the US), so it wouldn't be correct. If you are making a derived implementation, then it would be necessary to copy the while BSD 3-clause (you can find other places in the code where we've done this). Otherwise, it's simply respectful to include an acknowledgement in a comment – similar to how you would cite any other paper.

mbauman committed:
fix unalias support for empty arrays

mbauman commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

ararslan commented:
Looks like Nanosoldier ignored you. @nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25890#issuecomment-366098085) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/3c77c08104a113022aac8f376b1dc6f2c724d7b5/15fcaab_vs_9ac2dfa/report.md). cc @ararslan

mbauman committed:
Major simplification: do not worry about array extents

By defining `mightalias(::SubArray, ::SubArray)` we cover the vast majority of cases where extents are important. SubArrays are the basically only array in the standard library that "restrict" views into parents; all other wrappers largely re-shuffle the data. `dataids(::AbstractArray)` now just returns a tuple of `UInt`s that describe the identities of the mutable "blocks" that the array references.  By default, this is just `(objectid(A),)`, allowing all custom arrays to detect situations with two `===` arrays.  Anything more complicated requires them to define their component parts with a custom `dataids` method definition.

mbauman commented:
Ok.  I've pushed a third take on this, now.  Sorry for the successive re-designs here, but I'm much happier with the way this works now.

In short: `mightalias(A::AbstractArray, B::AbstractArray)` no longer worries about array extents.  It simply returns true if any of the data components in `A` are shared with any of the data components in `B`.  This is done by asking if their `dataids` overlap.

`dataids(A)` returns a tuple of `UInt`s; each `UInt` describes a memory region.  These are three of the definitions in this PR:

```
dataids(A::AbstractArray) = (objectid(A),)
dataids(A::Array) = pointer(A)
dataids(A::ReshapedArray) = dataids(A.parent)
```

This means that — by default — we'll detect aliasing custom arrays that are `===` (since their `objectid`s will match).  We'll also detect aliasing views of those arrays.  If, however, the custom array is a wrapper around another array and it doesn't define its own `dataids` function, then that is when `mightalias` will wrongly report `false`.

While this doesn't look at array extents anymore, a `mightalias(::SubArray, ::SubArray)` method covers the most common case with two views into the same parent.

Finally, no arrays should need to overload `unalias` anymore.  Instead, they can implement `copypreservingtype`.  Without a custom implementation, it falls back to `deepcopy`.  This is slow and overkill, but it's the best we got since we don't know how deep to go without a custom implementation.

--------------------

Some big takeaways from working through this:

* The `parent` function is totally worthless.  I tried really hard to come up with a design that would take advantage of its existence, but it's simply not helpful.  I'm even more convinced there's absolutely nothing useful you can do with `parent` without knowing what the array type is already.
* Using a numpy-like comparison of strided arrays is interesting, but I just don't think it's worth doing.  Given that the major consideration here is _wrapper_ arrays that typically define custom mappings from their indices into the parent indices, we could _ask them_ how to conservatively remap the indices. It needn't be exact. Then comparing a set of indices into the same parent is very quick and easy.  I started down that road, but it just got too messy for a first iteration when this is already a huge improvement.

andyferris commented on base/abstractarray.jl:
```diff
@@ -1048,6 +1048,86 @@ function _setindex!(::IndexCartesian, A::AbstractArray, v, I::Vararg{Int,M}) whe
+affecting the returned object. No guarantees are provided, and each custom array must
+opt-into aliasing detection by overloading this method by specializing on the second argument.
+
+This function must return an object of exactly the same type as `A` for performance and type-stability.
```
nitpick: no hyphen

mbauman committed:
Merge branch 'master' into mb/aliaasing

mbauman committed:
Remove hyphen in type-stability

mbauman commented:
Alright, I'd like to merge this guy tonight or tomorrow so I can then merge https://github.com/JuliaLang/julia/pull/24368.

vtjnash commented on base/abstractarray.jl:
```diff
@@ -1047,6 +1047,86 @@ function _setindex!(::IndexCartesian, A::AbstractArray, v, I::Vararg{Int,M}) whe
+unalias(dest, A) = mightalias(dest, A) ? copypreservingtype(A) : A
+
+copypreservingtype(A::Array) = copy(A)
+copypreservingtype(A::AbstractArray) = (@_noinline_meta; deepcopy(A)::typeof(A))
```
I'm not sure `deepcopy` is ever correct. Maybe better just to give a method error?

mbauman replied:
Ah, you're right, I didn't consider how this would change the identities of the values that get assigned into the array.  Perhaps we should just fall back to `copy(A)`.  It'll be type-unstable but this function exists to allow you to fix it.

mbauman committed:
Use `copy` with a method type assertion instead of `deepcopy`

This will do what we want in almost all cases. We only hit this method if `A` is aliasing a mutable array... which means that `A` is mutable as well, and thus in many cases it has defined an appropriate `similar(A)` method that returns the same type.

vchuravy commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

mbauman commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25890#issuecomment-367754076) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/96a820b492e77374296da804174ebc49fc566dae/6237ae0_vs_da98033/report.md). cc @ararslan

StefanKarpinski commented:
How fast do you want a garbage answer?

mbauman commented:
I should be able to fix this without too much trouble.  I had it working just fine a few commits ago, and the current design is even simpler.

mbauman committed:
Avoid comparing dataids in some obvious cases

mbauman commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

mbauman commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/25890#issuecomment-368084067):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(1)) [1]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/b55930f37562ceb44494b05ec1255b310e09ed08/0f90de2_vs_cca1261/report.md)
cc @ararslan

mbauman commented:
Maybe I'll just try again?

@nanosoldier `runbenchmarks(ALL, vs=":master")`

mbauman commented:
Maybe I'll just try again?

@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/25890#issuecomment-368120627):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(1)) [1]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/23ea2d278cc58ee46d1dd48d27a99b63cfd69f81/ee4d27c_vs_018fb61/report.md)
cc @ararslan

mbauman commented:
Same error — this is very strange.

```
ERROR: LoadError: MethodError: no method matching sqrt!(::Float64)
Closest candidates are:
  sqrt!(!Matched::AbstractArray) at statistics.jl:263
Stacktrace:
 [1] _std(::Array{Float64,1}, ::Bool, ::Nothing, ::Colon) at ./statistics.jl:298
 [2] #std#703 at ./statistics.jl:287 [inlined]
 [3] std at ./statistics.jl:287 [inlined]
 [4] perf_micro_randmatstat(::Int64) at /home/nanosoldier/.julia/v0.7/BaseBenchmarks/src/micro/methods.jl:109
```

I don't touch `sqrt!` here...

KristofferC commented:
But you do touch broadcasting, which is used in the stackframe above?

mbauman commented:
Turns out `std(rand(10))` was broken on master.  Whoops.  #26186

alyst commented on base/abstractarray.jl:
```diff
@@ -1047,6 +1047,88 @@ function _setindex!(::IndexCartesian, A::AbstractArray, v, I::Vararg{Int,M}) whe
+unalias(dest, A::AbstractRange) = A
+unalias(dest, A) = A
+
+copypreservingtype(A::Array) = copy(A)
```
This sounds like "copy-preserving type" and too similar to `eltype()`, `keytype()` etc.
Maybe something along `sametypecopy()`/`keeptypecopy()`/`unaliasingcopy()` (my fav)?

mbauman replied:
Fair point.  I also like `unaliasingcopy` for now — it's likely that we'll try to separate the two meanings of copy in the future.  By naming this more narrowly it gives us space for that design process.

mbauman commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25890#issuecomment-368242132) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/65982b0f2d8217b32defdd76c653d72465a64638/998a342_vs_e4b6677/report.md). cc @ararslan

mbauman committed:
Rename copypreservingtype to unaliascopy and update docs

mbauman commented:
Ok, the other few regressions here are all due to taking views of complicated non-`Array` types.  I want to ensure that modifying the parent array won't change any indices when we construct the view, but this one hurts because — unlike most other `unalias` calls — this one precedes an O(1) operation.  It's slow because we don't (yet) have specialized codegen for `objectid`.  It's fast for `Array` (and wrappers thereof) because we do have that specialized codegen support for `pointer`.  I'd advocate for merging this as it stands and dealing with the codegen of `objectid` separately — particularly since I won't be the one best suited to solve that issue.

I pushed a final (I hope) commit here that renames `copypreservingtype` to `unaliascopy` and updated all the docstrings.
